File: vb/src/Vanni.Idmix.Vb/RadixCodec.vb
Imports System.IO
Imports System.Numerics
Imports System.Text

Namespace Vanni.Idmix

''' <summary>XID 文本层：自定义进制编解码。</summary>
Public NotInheritable Class RadixCodec
    Private ReadOnly _base As Integer
    Private ReadOnly _chars As String
    Private ReadOnly _fromCustom As Dictionary(Of Char, Integer)

    Public Sub New(alphabet As String)
        If alphabet.Length < 2 Then
            Throw New ArgumentException("alphabet must have at least 2 unique characters", NameOf(alphabet))
        End If
        _base = alphabet.Length
        _chars = alphabet
        _fromCustom = New Dictionary(Of Char, Integer)()
        For i = 0 To alphabet.Length - 1
            Dim c = alphabet(i)
            If _fromCustom.ContainsKey(c) Then
                Throw New ArgumentException($"alphabet contains duplicate character {c}", NameOf(alphabet))
            End If
            _fromCustom(c) = i
        Next
    End Sub

    Public ReadOnly Property Base As Integer
        Get
            Return _base
        End Get
    End Property

    Public ReadOnly Property Chars As String
        Get
            Return _chars
        End Get
    End Property

    Public Function EncodeBytes(data As Byte()) As String
        If data.Length = 0 Then Return _chars(0).ToString()
        Dim buf(1 + data.Length) As Byte
        buf(0) = CByte((data.Length >> 8) And &HFF)
        buf(1) = CByte(data.Length And &HFF)
        Buffer.BlockCopy(data, 0, buf, 2, data.Length)
        Dim bytes = CType(buf.Clone(), Byte())
        If BitConverter.IsLittleEndian Then Array.Reverse(bytes)
        Dim n As New BigInteger(bytes)
        Return IntToString(n)
    End Function

    Public Function DecodeBytes(s As String) As Byte()
        If String.IsNullOrEmpty(s) Then Throw New ArgumentException("empty string", NameOf(s))
        Dim n = StringToInt(s)
        Dim raw = n.ToByteArray()
        If raw.Length > 0 AndAlso raw(raw.Length - 1) = 0 Then
            ReDim Preserve raw(raw.Length - 2)
        End If
        Array.Reverse(raw)
        For pad = 0 To 1
            Dim bufLen = pad + raw.Length
            Dim buf(bufLen - 1) As Byte
            If pad > 0 Then buf(0) = 0
            Buffer.BlockCopy(raw, 0, buf, pad, raw.Length)
            If buf.Length < 2 Then Continue For
            Dim dataLen = (buf(0) << 8) Or buf(1)
            If buf.Length <> 2 + dataLen Then Continue For
            Dim outBuf(dataLen - 1) As Byte
            Buffer.BlockCopy(buf, 2, outBuf, 0, dataLen)
            Return outBuf
        Next
        Throw New ArgumentException("invalid encoded data length")
    End Function

    Private Function IntToString(n As BigInteger) As String
        If n.IsZero Then Return _chars(0).ToString()
        Dim baseBi As New BigInteger(_base)
        Dim sb As New StringBuilder()
        While n > BigInteger.Zero
            Dim remVal As BigInteger
            n = BigInteger.DivRem(n, baseBi, remVal)
            sb.Append(_chars(CInt(remVal)))
        End While
        Dim chars = sb.ToString().ToCharArray()
        Array.Reverse(chars)
        Return New String(chars)
    End Function

    Private Function StringToInt(s As String) As BigInteger
        Dim n = BigInteger.Zero
        Dim baseBi As New BigInteger(_base)
        For Each c In s
            If Not _fromCustom.ContainsKey(c) Then
                Throw New ArgumentException($"invalid character {c}", NameOf(s))
            End If
            n = n * baseBi + _fromCustom(c)
        Next
        Return n
    End Function
End Class

End Namespace

File: csharp/src/Vanni.Idmix/RadixCodec.cs
using System.Numerics;
using System.Text;

namespace Vanni.Idmix;

/// <summary>基于自定义字符表的 Base-N 编解码器（默认 idmix 文本层）。</summary>
public sealed class RadixCodec : ICodec
{
    private readonly int _base;
    private readonly string _chars;
    private readonly Dictionary<char, int> _fromCustom;

    public RadixCodec(string alphabet)
    {
        if (alphabet.Length < 2)
            throw new ArgumentException("alphabet must have at least 2 unique characters", nameof(alphabet));
        _base = alphabet.Length;
        _chars = alphabet;
        _fromCustom = new Dictionary<char, int>();
        for (var i = 0; i < alphabet.Length; i++)
        {
            var c = alphabet[i];
            if (_fromCustom.ContainsKey(c))
                throw new ArgumentException($"alphabet contains duplicate character {c}", nameof(alphabet));
            _fromCustom[c] = i;
        }
    }

    public string Alphabet => _chars;
    public int Base => _base;

    public string Encode(byte[] data)
    {
        if (data.Length == 0) return _chars[0].ToString();
        var buf = new byte[2 + data.Length];
        buf[0] = (byte)((data.Length >> 8) & 0xFF);
        buf[1] = (byte)(data.Length & 0xFF);
        Buffer.BlockCopy(data, 0, buf, 2, data.Length);
        var n = new BigInteger(buf, isUnsigned: true, isBigEndian: true);
        return IntToString(n);
    }

    public byte[] Decode(string s)
    {
        if (string.IsNullOrEmpty(s))
            throw new ArgumentException("empty string", nameof(s));
        var n = StringToInt(s);
        var raw = n.ToByteArray(isUnsigned: true, isBigEndian: true);
        for (var pad = 0; pad <= 1; pad++)
        {
            var buf = new byte[pad + raw.Length];
            Buffer.BlockCopy(raw, 0, buf, pad, raw.Length);
            if (buf.Length < 2) continue;
            var dataLen = (buf[0] << 8) | buf[1];
            if (buf.Length != 2 + dataLen) continue;
            var outBuf = new byte[dataLen];
            Buffer.BlockCopy(buf, 2, outBuf, 0, dataLen);
            return outBuf;
        }
        throw new ArgumentException("invalid encoded data length");
    }

    private string IntToString(BigInteger n)
    {
        if (n.IsZero) return _chars[0].ToString();
        var baseBi = new BigInteger(_base);
        var sb = new StringBuilder();
        while (n > BigInteger.Zero)
        {
            n = BigInteger.DivRem(n, baseBi, out var rem);
            sb.Append(_chars[(int)rem]);
        }
        var chars = sb.ToString().ToCharArray();
        Array.Reverse(chars);
        return new string(chars);
    }

    private BigInteger StringToInt(string s)
    {
        var n = BigInteger.Zero;
        var baseBi = new BigInteger(_base);
        foreach (var c in s)
        {
            if (!_fromCustom.TryGetValue(c, out var idx))
                throw new ArgumentException($"invalid character {c}", nameof(s));
            n = n * baseBi + idx;
        }
        return n;
    }
}
